src/effects/VST3/VST3Effect.cpp

```diff
@@ -1217,3 +1183,128 @@ void VST3Effect::ReloadUserOptions()
 
    SetBlockSize(mUserBlockSize);
 }
+
+
+bool VST3Effect::TransferDataToWindow(const EffectSettings& settings)
+{
+   // Update parameter values in the plugin
+   if (StoreSettings(GetSettings(settings)))
+   {
+      return true;
+   }
+   
+   return false;
+}
+
+
+
+bool VST3Wrapper::ForEachParameter(ParameterVisitor visitor) const
+{
+   if (mEditController == nullptr)
+      return false;
+
+   for (int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      ParameterInfo pi;
+
+      Steinberg::Vst::ParameterInfo vstParameterInfo{ };
+      if (mEditController->getParameterInfo(i, vstParameterInfo) == Steinberg::kResultOk)
+      {
+         if (vstParameterInfo.flags & Steinberg::Vst::ParameterInfo::kCanAutomate)
+         {
+            if (!visitor(ParameterInfo{ vstParameterInfo }))
+               break;
+         }
+      }
+   }
+
+   return true;
+}
+
+
+bool VST3Wrapper::LoadPreset(Steinberg::IBStream* fileStream, const Steinberg::FUID& classID)
+{
+   using namespace Steinberg;
+
+   return Vst::PresetFile::loadPreset
+   (
+      fileStream,
+      FUID::fromTUID(classID),
+      mEffectComponent.get(),
+      mEditController.get()
+   );
+}
+
+
+bool VST3Wrapper::SavePreset(Steinberg::IBStream* fileStream, const Steinberg::FUID& classID) const
+{
+   using namespace Steinberg;
+
+   return Vst::PresetFile::savePreset
+   (
+      fileStream,
+      classID,
+      mEffectComponent.get(),
+      mEditController.get()
+   );
+}
+
+
+bool VST3Wrapper::FetchSettings(VST3EffectSettings& settings) const
+{
+   settings.mValues.clear();
+
+   return ForEachParameter(
+
+      [&](const ParameterInfo& pi)
+   {
+      const auto id = pi.id;
+      settings.mValues[id] = mEditController->getParamNormalized(id);
+      return true;
+   }
+   );
+}
+
+
+bool VST3Wrapper::StoreSettings(const VST3EffectSettings& settings) const
+{
+   return ForEachParameter(
+
+      [&](const ParameterInfo& pi)
+      {
+         if (mComponentHandler == nullptr)
```

Comment:
The component handler operations should not be in this function; but `setParamNormalized` needs to be called, so there should be a null check on the edit controller.